content/post/2024-03-12-mocking/index.Rmd

```diff
@@ -0,0 +1,143 @@
+---
+slug: mocking-new-take
+title: "Update on mocking for testing R packages" 
+authors: 
+- Maëlle Salmon
+date: "2024-03-13" 
+tags: 
+- package development
+- testing
+- mocking
+output: hugodown::hugo_document
+---
+
+```{r setup, include=FALSE}
+
+knitr::opts_chunk$set(fig.path = "", comment = "")
+# knitr hook to make images output use Hugo options
+knitr::knit_hooks$set(
+  plot = function(x, options) {
+    hugoopts <- options$hugoopts
+    paste0(
+      "{{<figure src=",
+      '"', x, '" ',
+      if (!is.null(hugoopts)) {
+        glue::glue_collapse(
+          glue::glue('{names(hugoopts)}="{hugoopts}"'),
+          sep = " "
+        )
+      },
+      ">}}\n"
+    )
+  }
+)
+
+```
+
+```{r, include = FALSE, cache = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>",
+  out.width = "100%",
+  asciicast_theme = "pkgdown"
+)
+```
+
+This blog featured a [post on mocking, the art of replacing a function with whatever fake we need for testing](/2019/10/29/mocking/), years ago.
+Since then, we've entered a new decade, the second edition of Hadley Wickham's and Jenny Bryan's [R packages book](https://r-pkgs.org) was published, and mocking returned to testthat, so it's time for a new take/resources roundup!
+
+## Mocking yay or nay
+
+The R packages book by Hadley Wickham and Jenny Bryan contains an insightful paragraph on [test coverage](https://r-pkgs.org/testing-design.html#sec-testing-design-coverage)[^1], including those lines:
+
+[^1]: The three chapters on testing are definitely worth a read!
+
+> "In many cases, that last 10% or 1% often requires some awkward gymnastics to cover. Sometimes this forces you to introduce mocking or some other new complexity. Don't sacrifice the maintainability of your test suite in the name of covering some weird edge case that hasn't yet proven to be a problem."
+
+This is definitely good to keep in mind whilst diving into the world of mocking.
+Not that mocking needs to be especially complex!
+
+Beside, sometimes, testing is complex because your code is begging to be refactored.
+:wink:
+
+## Revisiting the general mocking example
+
+The [example of general mocking](/2019/10/29/mocking/#general-mocking) in the post from 2019 can be rewritten to take advantage of the new mocking functionality that was [announced as no longer experimental in October 2023](https://www.tidyverse.org/blog/2023/10/testthat-3-2-0/#mocking).
```

Comment:
Is the passive voice here a conscious decision?

Reply:
nope, fixed!